cmd/vcfFilter/expression.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"github.com/vertgenlab/gonomics/vcf"
+	"log"
+	"strconv"
+	"strings"
+)
+
+type operator string
+
+const (
+	equal        operator = "="
+	notEqual     operator = "!="
+	greater      operator = ">"
+	less         operator = "<"
+	greaterEqual operator = ">="
+	lessEqual    operator = "<="
+	present      operator = ""
+)
+
+// TODO complex expression parsing
+// parseExpression parses a ';' delimited string to a slice of boolean functions to test a vcf record.
+func parseExpression(input string, header vcf.Header, isFormatElseInfo bool, includeMissingInfo bool) testingFuncs {
+	var answer testingFuncs
+	input = strings.Trim(input, "\"") // trim quotations from the ends
+	expSlice := strings.Split(input, ";")
+	for _, exp := range expSlice {
+		var op operator
+		var tagValuePair []string
+		var tag, value string
+		op = searchOp(exp)
+		if op != present {
+			tagValuePair = strings.Split(exp, string(op))
+		} else {
+			tagValuePair = []string{exp}
+		}
+		tag = strings.Trim(tagValuePair[0], " ")
+		if len(tagValuePair) == 2 {
+			value = strings.Trim(tagValuePair[1], " ")
+		}
+		answer = append(answer, getRelationshipTest(tag, value, op, header, isFormatElseInfo, includeMissingInfo))
+	}
+	return answer
+}
+
+func searchOp(exp string) operator {
+	if strings.Contains(exp, ">=") {
+		return greaterEqual
+	}
+
+	if strings.Contains(exp, "<=") {
+		return lessEqual
+	}
+
+	if strings.Contains(exp, "!=") {
+		return notEqual
+	}
+
+	if strings.Contains(exp, "=") {
+		return equal
+	}
+
+	if strings.Contains(exp, ">") {
+		return greater
+	}
+
+	if strings.Contains(exp, "<") {
+		return less
+	}
+
+	return present
+}
+
+func getRelationshipTest(tag string, value string, r operator, header vcf.Header, isFormatElseInfo bool, includeMissingInfo bool) func(vcf.Vcf) bool {
+	var tagKey vcf.Key
+	if isFormatElseInfo {
+		tagKey = header.Format[tag].Key
+	} else {
+		tagKey = header.Info[tag].Key
+	}
+
+	if tagKey.Number != "1" && tagKey.Number != "0" {
+		log.Printf("WARNING: expressions for tags with multiple values will be true if any value passes filter. Tag '%s' has '%s' fields.", tag, tagKey.Number)
+	}
+
+	switch tagKey.DataType {
+	case vcf.Integer:
+		test := r.TestInteger()
+		val, err := strconv.Atoi(value)
+		if err != nil {
+			log.Fatalf("Error: value '%s' is not an integer as expected for tag '%s'.", value, tag)
+		}
+		return func(v vcf.Vcf) bool {
+			var answer bool = true
+			queryResult, found := vcf.QueryInt(v, tagKey)
+			if !found {
+				return includeMissingInfo
+			}
+			for _, recordVal := range queryResult[0] {
```

Comment:
I think that here we are looping over all values of the tag and since each needs to pass, we are doing an "and" over those tests.  I think this is probably the most useful test, but with only a few (small?) changes I think we could pass a flag that would do an "or" over them.